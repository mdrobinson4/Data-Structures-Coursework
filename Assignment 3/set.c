#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <assert.h>
#include "set.h"
#include "da.h"
#include "real.h"

static int link(SET *s1, SET *s2);
static SET *getRep(SET *s);

struct set {
  DA *store;
  void *value;
  SET *rep;
  SET *parent;
  int rank;
  int index;
  void (*display)(void *, FILE *);
  void (*free)(void *);
};

SET *newSET(void) {
  SET *s = malloc(sizeof(SET));
  assert(s != 0);
  s->store = newDA();
  s->display = 0;
  s->free = 0;
  s->rep = s;
  s->parent = s;
  s->rank = 0;
  s->index = 0;
  return s;
}

void setSETdisplay(SET *s,void (*d)(void *,FILE *)) {
  setDAdisplay(s->store, d);
  s->display = d;
}

void setSETfree(SET *s,void (*d)(void *)) {
  setDAfree(s->store, d);
  s->free = d;
}

int makeSET(SET *d, void *value) {
  SET *s = newSET();
  s->value = value;
  s->index = sizeDA(d->store);
  s->parent = s;
  insertDAback(d->store, s);
  return s->index;
}

int findSET(SET *d, int index) {
  SET *s = getDA(d->store, index);
  s = getRep(s);
  return s->index;
}

SET *getRep(SET *s) {
  if (s != s->parent)
    s->parent = getRep(s->parent);
  return s->parent;
}

int unionSET(SET *d,int index1,int index2) {
  int repIndex =  link(getDA(d->store, findSET(d, index1)), getDA(d->store, findSET(d, index2)));
  d->rep->index = repIndex;
  return repIndex;
}

int link(SET *s1, SET *s2) {
  if (s1->rank > s2->rank) {
    s2->parent = s1;
    return s1->index;
  }
  if (s2->rank > s1->rank) {
    s1->parent = s2;
    return s2->index;
  }
  if (s1->index < s2->index) {
    s2->parent = s1;
    s1->rank += 1;
    return s1->index;
  }
  if (s1->index > s2->index) {
    s1->parent = s2;
    s2->rank += 1;
    return s2->index;
  }
  return s1->index;
}

void displaySET(SET *d, FILE *fp) {
  for (int i = 0; i < sizeDA(d->store); i++) {
    SET *s = getDA(d->store, i);
    fprintf(fp, "%d: ", i);
    if (s == s->parent) {
      d->display(s->value, fp);
      fprintf(fp, "(%d)\n", s->rank);
    }
    else {
      while (s != s->parent) {
        d->display(s->value, fp);
        s = s->parent;
        fprintf(fp, " ");
      }
      d->display(s->value, fp);
      fprintf(fp, "\n");
    }
  }
}

int debugSET(SET *d,int level) {
  debugDA(d->store, level);
  return 0;
}

void freeSET(SET *d) {
  for (int i = 0; i < sizeDA(d->store); i++) {
    SET *s = getDA(d->store, i);
    d->free(s->value);
    setDAfree(s->store, 0);
    freeDA(s->store);
    free(s);
  }
  setDAfree(d->store, 0);
  freeDA(d->store);
  free(d);
}
