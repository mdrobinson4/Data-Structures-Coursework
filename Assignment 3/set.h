#ifndef __SET_INCLUDED__
#define __SET_INCLUDED__

#include <stdio.h>

typedef struct set SET;

extern SET  *newSET(void);
extern void  setSETdisplay(SET *s,void (*d)(void *,FILE *));
extern void  setSETfree(SET *s,void (*d)(void *));
extern int   makeSET(SET *d,void *value);
extern int   findSET(SET *d,int index);
extern int   unionSET(SET *d,int index1,int index2);
extern void  displaySET(SET *d,FILE *fp);
extern int   debugSET(SET *d,int level);
extern void  freeSET(SET *d);

#endif
