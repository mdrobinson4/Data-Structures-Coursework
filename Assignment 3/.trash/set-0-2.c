#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "integer.h"
#include "real.h"
#include "string.h"
#include "set.h"

void srandom(unsigned int);
long int random(void);

int
main(void)
    {
    srandom(3);
    printf("REAL test of SET, simple\n");
    int i,size=10,count=8;
    SET *p = newSET();
    setSETdisplay(p,displayREAL);
    setSETfree(p,freeREAL);
    int *spots = malloc(sizeof(int) * size);
    REAL **elements = malloc(sizeof(REAL *) * size);
    assert(spots != 0 && elements != 0);
    for (i = 0; i < size; ++i)
        {
        spots[i] = -1;
        elements[i] = newREAL(i * 2.0);
        }
    for (i = 0; i < size; ++i)
        {
        int j = random() % size;
        REAL *temp = elements[i];
        elements[i] = elements[j];
        elements[j] = temp;
        }
    for (i = 0; i < size; ++i)
        {
        spots[i] = makeSET(p,elements[i]);
        }
    displaySET(p,stdout);
    for (i = 0; i < count; ++i)
        {
        int a = random() % size;
        int b = random() % size;
        printf("unioning %d and %d\n",a,b);
        unionSET(p,a,b);
        }
    displaySET(p,stdout);
    freeSET(p);
    free(spots);
    free(elements);
    return 0;
    }
