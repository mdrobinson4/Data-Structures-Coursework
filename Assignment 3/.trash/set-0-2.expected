REAL test of SET, simple
0: 8.000000(0)
1: 4.000000(0)
2: 0.000000(0)
3: 12.000000(0)
4: 18.000000(0)
5: 6.000000(0)
6: 14.000000(0)
7: 16.000000(0)
8: 10.000000(0)
9: 2.000000(0)
unioning 8 and 4
unioning 3 and 9
unioning 2 and 1
unioning 4 and 5
unioning 1 and 5
unioning 5 and 2
unioning 8 and 8
unioning 3 and 7
0: 8.000000(0)
1: 4.000000(2)
2: 0.000000 4.000000
3: 12.000000(1)
4: 18.000000 4.000000
5: 6.000000 4.000000
6: 14.000000(0)
7: 16.000000 12.000000
8: 10.000000 4.000000
9: 2.000000 12.000000
